appdatabase/migrations/sql/1683627613_accounts_and_keycards_improvements.up.sql

```diff
@@ -0,0 +1,94 @@
+-- Possible `operable` column values:
+-- "no"        // an account is non operable it is not a keycard account and there is no keystore file for it and no keystore file for the address it is derived from
+-- "partially" // an account is partially operable if it is not a keycard account and there is created keystore file for the address it is derived from
+-- "fully"     // an account is fully operable if it is not a keycard account and there is a keystore file for it
+
+-- Adding new tables `keypairs` and `keypairs_accounts`
+CREATE TABLE IF NOT EXISTS keypairs (
+  key_uid VARCHAR PRIMARY KEY NOT NULL CHECK (length(trim(key_uid)) > 0),
+  name VARCHAR NOT NULL DEFAULT "",
+  type VARCHAR NOT NULL DEFAULT "",
+  derived_from VARCHAR NOT NULL DEFAULT "",
+  last_used_derivation_index INT NOT NULL DEFAULT 0,
+  synced_from VARCHAR NOT NULL DEFAULT "", -- keeps an info which device this keypair is added from
+  clock INT NOT NULL DEFAULT 0
+);
+
+CREATE TABLE IF NOT EXISTS keypairs_accounts (
+  address VARCHAR PRIMARY KEY,
+  key_uid VARCHAR,
+  pubkey VARCHAR,
+  path VARCHAR NOT NULL DEFAULT "",
+  name VARCHAR NOT NULL DEFAULT "",
+  color VARCHAR NOT NULL DEFAULT "",
+  emoji VARCHAR NOT NULL DEFAULT "",
+  wallet BOOL NOT NULL DEFAULT FALSE,
+  chat BOOL NOT NULL DEFAULT FALSE,
+  hidden BOOL NOT NULL DEFAULT FALSE,
+  operable VARCHAR NOT NULL DEFAULT "no", -- describes an account's operability (read an explanation at the top of this file)
+  created_at DATETIME NOT NULL,
+  updated_at DATETIME NOT NULL,
+  clock INT NOT NULL DEFAULT 0,
+  FOREIGN KEY(key_uid) REFERENCES keypairs(key_uid) 
+    ON DELETE CASCADE
+);
+
+-- Fulfilling the tables
+INSERT INTO keypairs 
+  SELECT key_uid, keypair_name, "profile", derived_from, last_used_derivation_index, "", clock 
+  FROM accounts 
+  WHERE type != "watch" AND type != "seed" AND type != "key"
+  GROUP BY key_uid;
+
+INSERT INTO keypairs 
+  SELECT key_uid, keypair_name, type, derived_from, last_used_derivation_index, "", clock 
+  FROM accounts 
+  WHERE type != "watch" AND type != "" AND type != "generated"
```

Comment:
oh, just saw another fulfilling, so maybe these 2 inserts could combine into one and use my above condition?

Reply:
This one inserts all accounts referring to a seed or private key keypairs.

Reply:
> emm.. wouldn't it be much clear to use WHERE type = "" or type = "generated" or type = "seed" or type = "key"  instead of not condition?

No we cannot. In the old `accounts` table, accounts which were referring to "profile" had type either `""` (empty) in case of default wallet account and chat account or `generated` in case of accounts generated from the profile (master) keypair. In new table profile keypair `type` is set to `profile` and that's why we have the first insertion like this ` SELECT key_uid, keypair_name, "profile", derived_from,...` so `"profile"` is something we don't have where to get it from.

In second insertion, since we use `seed` for seed imported keypairs and `key` for private key imported keypairs, which are the same values as "old" account's type, we are able to use this sql `SELECT key_uid, keypair_name, type, derived_from,...` cause it refers to correct type values.

Hence that's where the conditions in both cases come from.

Reply:
the records for type="seed" and type="key" shouldn't be treated as 2 separate keypairs? because they may have different master key

Reply:
> the records for type="seed" and type="key" shouldn't be treated as 2 separate keypairs?

How do you mean "shouldn't"?
They are different and they refer to the way how a user imported new keypair, via seed phrase or via private key. 

If you have seed phrase and private key for the same master key, meaning that they both generate the same `keyUid`, you will be able to add that keypair only once either via seed phrase or private key (or you can add it via seed phrase then delete it and add the same keypair via private key), but you definitely cannot add the same master key twice, once via seed phrase once via private key.

Reply:
let's consider following user cases:
after logined status mobile app, user add a wallet account(import from seed) A using seed phrase "a b c..." , then add a wallet account(import from seed) B using seed phrase "d e f ...", then add wallet account(import from pk) C using private key "ghi", then add wallet account(import from pk) D using private key "jkl", what/how many records we will expect for table keypairs ?

Reply:
So after user logged in:
- that means there is already a `profile` keypair added to `keypairs` table (this is first keypair in `keypairs` table)
- `user add a wallet account(import from seed) A using seed phrase "a b c..."` this is second keypair in `keypairs` table (but the first one with type `seed`) and will have `keyUid` = `abc123`
- `then add a wallet account(import from seed) B using seed phrase "d e f ..."` this is third keypair in `keypairs` table (but the second one with type `seed`) and will have `keyUid` = `def123`
- `then add wallet account(import from pk) C using private key "ghi"` this will be fourth keypair in `keypairs` table (but the first one with type `key`) and will have `keyUid` = `ghi123`
- `then add wallet account(import from pk) D using private key "jkl"` this will be fifth keypair in `keypairs` table (but the second one with type `key`) and will have `keyUid` = `jkl123`

> what/how many records we would expect for table keypairs ?

`keypairs` table will have 5 rows

Reply:
I see, thank you for your explain, surprise for me 😂 . I thought we still use the same keyUid for different wallet account as used to be ... and one keyUid means one login account